runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingModeExecutionContext.java

```diff
@@ -581,30 +594,96 @@ public void flushState() {
-                .filter(
-                    timer ->
-                        WindmillTimerInternals.isUserTimer(timer)
-                            && timer.getStateFamily().equals(stateFamily))
-                .transform(
-                    timer ->
-                        WindmillTimerInternals.windmillTimerToTimerData(
-                            WindmillNamespacePrefix.USER_NAMESPACE_PREFIX, timer, windowCoder))
-                .iterator();
+            Iterators.peekingIterator(
+                FluentIterable.from(StreamingModeExecutionContext.this.getFiredTimers())
+                    .filter(
+                        timer ->
+                            WindmillTimerInternals.isUserTimer(timer)
+                                && timer.getStateFamily().equals(stateFamily))
+                    .transform(
+                        timer ->
+                            WindmillTimerInternals.windmillTimerToTimerData(
+                                WindmillNamespacePrefix.USER_NAMESPACE_PREFIX, timer, windowCoder))
+                    .iterator());
       }
 
-      if (!cachedFiredUserTimers.hasNext()) {
-        return null;
+      while (cachedFiredUserTimers.hasNext()) {
+        TimerData nextInBundle = cachedFiredUserTimers.peek();
+        NavigableSet<TimerData> modifiedUserTimersOrdered =
+            getModifiedUserTimersOrdered(nextInBundle.getDomain());
+        // If there is a modified timer that is earlier than the next timer in the bundle, try and
+        // fire that first.
+        while (!modifiedUserTimersOrdered.isEmpty()
+            && modifiedUserTimersOrdered.first().compareTo(nextInBundle) <= 0) {
+          TimerData earlierTimer = modifiedUserTimersOrdered.pollFirst();
+          if (!timerModified(earlierTimer)) {
+            // We must delete the timer. This prevents it from being committed to the backing store.
+            // It also handles the
+            // case where the timer had been set to the far future and then modified in bundle;
+            // without deleting the
+            // timer, the runner will still have that future timer stored, and would fire it
+            // spuriously.
+            userTimerInternals.deleteTimer(earlierTimer);
+            return earlierTimer;
+          }
+        }
+        // There is no earlier timer to fire, so return the next timer in the bundle.
+        nextInBundle = cachedFiredUserTimers.next();
+        if (!timerModified(nextInBundle)) {
+          // User timers must be explicitly deleted when delivered, to release the implied hold.
+          userTimerInternals.deleteTimer(nextInBundle);
+          return nextInBundle;
+        }
       }
-      TimerData nextTimer = cachedFiredUserTimers.next();
-      // User timers must be explicitly deleted when delivered, to release the implied hold
-      userTimerInternals.deleteTimer(nextTimer);
-      return nextTimer;
+      return null;
```

Comment:
Change return value to `@Nullable` (if this is really necessary)

Reply:
This isn't new - null was returned previously as well. Annotating here will require expanding this PR far beyond this file (GroupAlsoByWindows, SImpleParDoFn, etc.) so seems more appropriate for another PR